tests/vm_test.go

```diff
@@ -1346,61 +1343,55 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
-						v1.KeepLauncherAfterFailureAnnotation: keepLauncher,
-					}
+					By("creating a Running VM")
+					vm := libvmi.NewVirtualMachine(libvmifact.NewGuestless(
+						libvmi.WithAnnotation(v1.KeepLauncherAfterFailureAnnotation, keepLauncher),
+						libvmi.WithAnnotation(v1.FuncTestLauncherFailFastAnnotation, ""),
+					), libvmi.WithRunStrategy(v1.RunStrategyOnce))
+
 					vm, err := virtClient.VirtualMachine(testsuite.GetTestNamespace(vm)).Create(context.Background(), vm, metav1.CreateOptions{})
 					Expect(err).ToNot(HaveOccurred())
 
-					By("Starting the VMI with virtctl")
-					startCommand := clientcmd.NewRepeatableVirtctlCommand(virtctl.COMMAND_START, "--namespace", vm.Namespace, vm.Name)
-					Expect(startCommand()).To(Succeed())
-
-					By("Waiting for VM to be in Starting status")
-					Eventually(ThisVM(vm), 160*time.Second, 1*time.Second).Should(HavePrintableStatus(v1.VirtualMachineStatusStarting))
-
 					By("Waiting for VMI to fail")
 					Eventually(ThisVMIWith(vm.Namespace, vm.Name), 480*time.Second, 1*time.Second).Should(BeInPhase(v1.Failed))
+					vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+					Expect(err).ShouldNot(HaveOccurred())
+
+					getComputeContainerStateRunning := func() (*k8sv1.ContainerStateRunning, error) {
+						launcherPod, err := libpod.GetPodByVirtualMachineInstance(vmi, vm.Namespace)
+						if err != nil {
+							return nil, err
+						}
+						for _, status := range launcherPod.Status.ContainerStatuses {
+							if status.Name == "compute" {
+								return status.State.Running, nil
+							}
+						}
+						return nil, nil
+					}
 
 					// If the annotation v1.KeepLauncherAfterFailureAnnotation is set to true, the containerStatus of the
 					// compute container of the virt-launcher pod is kept in the running state.
 					// If the annotation v1.KeepLauncherAfterFailureAnnotation is set to false or not set, the virt-launcher pod will become failed.
 					By("Verify that the virt-launcher pod or its container is in the expected state")
-					vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
-					Expect(err).ToNot(HaveOccurred())
-					launcherPod, err := libpod.GetPodByVirtualMachineInstance(vmi, vm.Namespace)
-					Expect(err).ToNot(HaveOccurred())
-
 					if toKeep, _ := strconv.ParseBool(keepLauncher); toKeep {
-						Consistently(func() bool {
-							for _, status := range launcherPod.Status.ContainerStatuses {
-								if status.Name == "compute" && status.State.Running != nil {
-									return true
-								}
-							}
-							return false
-						}, 10*time.Second, 1*time.Second).Should(BeTrue())
+						Consistently(func() *k8sv1.ContainerStateRunning {
+							computeContainerStateRunning, err := getComputeContainerStateRunning()
+							Expect(err).ToNot(HaveOccurred())
+							return computeContainerStateRunning
```

Comment:
You can rewrite this to `Consistently(getComputeContainerStateRunning)....` and it will just work.

Reply:
I agree that it would be simple but please see:
https://github.com/kubevirt/kubevirt/pull/12040#discussion_r1660912853

Reply:
I won't insist on this one

Reply:
I don't think that:
```
Consistently(getComputeContainerStateRunning).WithTimeout(10 * time.Second).WithPolling(1 * time.Second).ShouldNot(BeNil())
```
goes against https://github.com/kubevirt/kubevirt/pull/12040#discussion_r1660912853.
On the contrary, it satisfies it.
IIUIC the main issue is that we should not perform plain `Expect` inside `Eventually` or `Consistently` blocks, because the context of the `Expect` is different from the `Consistently` one.
If you want to use `Expect` you should need to pass the `g Gomega` to the function.

Consider that Gomega:
```
If multiple values are returned by the function, Eventually will 
pass the first value to the matcher and require that all others 
are zero-valued. This allows you to pass Eventuallya function 
that returns a value and an error - a common pattern in Go.
```
https://onsi.github.io/gomega/#category-2-making-codeeventuallycode-assertions-on-functions

So in this case, Lubo's suggestion is the right way to use the `Consistently`.

@nunnatsa Please chime in if this is wrong

Reply:
There are two reasons I don't like gomega helpers:
1. The error report is confusing. It's never clear where the error came from. If you use offset, then you don't know where exactly in the helper function you got assertion error. and if you don't use offset, then the error report points to the assertion failure, but you don't know which call failed, if you have multiple calls to the helper function.
2. I think assertion is part of the test, and the test should tell a story that you should read from stat to its end. Using helpers breaks this concept in a way. Also, the go way is to return error and check them. I think that helpers should return errors, and the test should assert these errors.

Reply:
Also, hiding `Expect` within a helper, does not solves the issue with the `Expect` in eventually. It just makes it harder to debug the issue. For example, I wrote this test:

```go
var _ = Describe("Checkginkgo", func() {

	i := 0
	testingSomething := func() error {
		i++
		Expect(i).To(BeNumerically(">", 5))
		return nil
	}

	It("should count", func() {
		Eventually(testingSomething).WithTimeout(10 * time.Second).WithPolling(time.Second).Should(Succeed())
	})
})
```

Here the test assert with `Expect`, without the gomega variable, from a helper function.

If the assumption above were right, then the test should succeed after 5 seconds, but it failed immediately with
```
  [FAILED] Expected
      <int>: 1
  to be >
      <int>: 5
  In [It] at: /home/nunnatsa/GolandProjects/checkginkgo/checkginkgo_suite_test.go:21 @ 09/08/24 15:43:21.101
```

Reply:
I do get the expected result when I'm doing
```go
var _ = Describe("Checkginkgo", func() {
	i := 0
	testingSomething := func(g Gomega) error {
		i++
		g.Expect(i).To(BeNumerically(">", 5))
		return nil
	}

	It("should count", func() {
		Eventually(testingSomething).
			WithTimeout(10 * time.Second).
			WithPolling(time.Second).
			Should(Succeed())
	})
})
```

Reply:
I understand that and I agree. I think we are saying the same thing: "Avoid using `Expect` inside the `Eventually`/`Consistently` loop. Or if you want to it, pass the `Gomega` variable and use it".
The thing is that here I also think that it is not necessary to use the `g.Expect` since we can delegate the check to the `Consistently`.
With the example you provided I am suggesting:
```go
var _ = Describe("Checkginkgo", func() {

	i := 0
	testingSomething := func() int {
		i++
		return i
	}

	It("should count", func() {
		Eventually(testingSomething).
			WithTimeout(10 * time.Second).
			WithPolling(time.Second).
			Should(BeNumerically(">", 5))
	})
})
```

And also, using the error check property of the `Consistently`:
```go
var _ = Describe("Checkginkgo", func() {

	i := 0
	testingSomething := func() (int, error) {
		i++
		if i == 2 {
			return 0, fmt.Errorf("generic")
		}
		return i, nil
	}

	It("should count", func() {
		Eventually(testingSomething).
			WithTimeout(10 * time.Second).
			WithPolling(time.Second).
			Should(BeNumerically(">", 5))
	})
})
```
This will pass as well

Reply:
@nunnatsa The helper in this case doesn't do asserts. Please read Federico's comments

Reply:
Hi @fossedihelm @xpivarc @nunnatsa , I think that @fossedihelm is a nice workaround, but it could be a bit confusing. In the example:

```
Eventually(testingSomething).
	WithTimeout(10 * time.Second).
	WithPolling(time.Second).
	Should(BeNumerically(">", 5))
```
It seems like we're only checking the int, so it might be hard to tell that an error was also returned from the second output variable of `func testingSomething`. That’s why I think the current approach might make it easier to catch the error. However, if there’s strong disagreement, I’m happy to adjust it. Just let me know! :)

Reply:
@Barakmor1 - it will work. Gomega will also check that the last returned value is nil, but I always prefer explicit approach. I don't like magics.